patches/api/0419-Paper-Plugins.patch

```diff
@@ -0,0 +1,2123 @@
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java b/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..fa3fc912272285b05352d69481c6eb3f8b89ea55
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
+@@ -0,0 +1,176 @@
++package io.papermc.paper.plugin.configuration;
++
++import org.bukkit.permissions.Permission;
++import org.bukkit.permissions.PermissionDefault;
++import org.bukkit.plugin.PluginLoadOrder;
++import org.bukkit.plugin.java.JavaPlugin;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.List;
++
++/**
++ * This class acts as an abstraction for a plugin configuration.
++ */
++@ApiStatus.NonExtendable
++public interface PluginMeta {
++
++    /**
++     * Gives the name of the plugin. This name is a unique identifier for
++     * plugins.
++     * <ul>
++     * <li>Will only contain alphanumeric characters, underscores, hyphon,
++     *     and period (a-z,A-Z,0-9,_.-).
++     * <li>Typically used for identifying the plugin data folder.
++     * <li>The is the token referenced in {@link #getDepend()}, {@link
++     *     #getSoftDepend()}, and {@link #getLoadBefore()}.
++     * </ul>
++     * <p>
++     * In the plugin.yml, this entry is named <code>name</code>.
++     * <p>
++     * Example:<blockquote><pre>name: MyPlugin</pre></blockquote>
++     *
++     * @return the name of the plugin
++     */
++    @NotNull
++    String getName();
++
++    /**
++     * Returns the name of a plugin, including the version.
++     *
++     * @return a descriptive name of the plugin and respective version
++     */
++    @NotNull
++    default String getDisplayName() {
++        return this.getName() + " v" + this.getVersion();
++    }
++
++    /**
++     * Gives the fully qualified name of the main class for a plugin.
++     * A {@link JavaPlugin} is expected at this location.
++     *
++     * @return the fully qualified main class for the plugin
++     */
++    @NotNull
++    String getMain();
++
++    /**
++     * Gives the phase of server startup that the plugin should be loaded.
++     *
++     * @return phase
++     */
++    @NotNull
++    PluginLoadOrder getLoad();
++
++    /**
++     * Gets the version of this plugin
++     *
++     * @return version string
++     */
++    @NotNull
++    String getVersion();
++
++    /**
++     * Gets the prefix that should be used for the plugin logger.
++     * Null if not specified
++     *
++     * @return logger prefix
++     */
++    @Nullable
++    String getPrefix();
++
++    /**
++     * Gets a list of dependencies that are required for this plugin to load.
++     *
++     * @return immutable list of required dependencies
++     */
++    @NotNull
++    List<String> getDepend();
```

Comment:
The line of dependency/provide methods should also be named something like `get(Plugin)Dependencies`, `getPluginsLoadingBefore` (as load before is ambiguous), `getProvidedPlugins`